ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,533 @@
+					loadAccountsAddResourceTask(nil, basics.CreatableIndex(fa), basics.AssetCreatable, task, resourceTasks, queue)
+				}
+				for ixa := range stxn.Txn.Accounts {
+					loadAccountsAddAccountTask(&stxn.Txn.Accounts[ixa], task, accountTasks, queue)
+				}
+			case protocol.CompactCertTx:
+				fallthrough
+			case protocol.KeyRegistrationTx:
+				fallthrough
+			default:
+			}
+			// If you add new addresses here, also add them in getTxnAddresses().
+			if !stxn.Txn.Sender.IsZero() {
+				loadAccountsAddAccountTask(&stxn.Txn.Sender, task, accountTasks, queue)
+			}
+		}
+		totalBalances += task.balancesCount
+		totalResources += task.resourcesCount
+		// expand the queue if needed.
+		queue = queue.expand()
+	}
+
+	// find the number of tasks
+	tasksCount := int64(0)
+	for lastQueueEntry := &tasksQueue; ; lastQueueEntry = lastQueueEntry.next {
+		if lastQueueEntry.next == nil {
+			tasksCount = int64(lastQueueEntry.baseIdx + lastQueueEntry.used)
+			break
+		}
+	}
+
+	// update all the groups task :
+	// allocate the correct number of balances, as well as
+	// enough space on the "done" channel.
+	allBalances := make([]loadedAccountDataEntry, totalBalances)
+	allResources := make([]loadedResourcesEntry, totalResources)
+	usedBalances := 0
+	usedResources := 0
+
+	// groupDoneCh is used to communicate the completion signal for a single
+	// resource/address load between the go-routines and the main output channel
+	// writer loop. The various go-routines would write to the channel the index
+	// of the task that is complete and ready to be sent.
+	groupDoneCh := make(chan groupTaskDone, len(groupsReady))
+	const dependencyFreeGroup = -int64(^uint64(0)/2) - 1
+	for grpIdx := range groupsReady {
+		gr := &groupsReady[grpIdx]
+		gr.groupTaskIndex = grpIdx
+		gr.incompleteCount = int64(gr.balancesCount + gr.resourcesCount)
+		gr.balances = allBalances[usedBalances : usedBalances+gr.balancesCount]
+		if gr.resourcesCount > 0 {
+			gr.resources = allResources[usedResources : usedResources+gr.resourcesCount]
+			usedResources += gr.resourcesCount
+		}
+		usedBalances += gr.balancesCount
+		if gr.incompleteCount == 0 {
+			gr.incompleteCount = dependencyFreeGroup
+		}
+	}
+
+	taskIdx := int64(-1)
+	defer atomic.StoreInt64(&taskIdx, tasksCount)
+	// create few go-routines to load asyncroniously the account data.
+	for i := 0; i < asyncAccountLoadingThreadCount; i++ {
+		go p.asyncPrefetchRoutine(&tasksQueue, &taskIdx, groupDoneCh)
+	}
+
+	// iterate on the transaction groups tasks. This array retains the original order.
+	completed := make(map[int]bool)
+	for i := 0; i < len(p.groups); {
+	wait:
+		incompleteCount := atomic.LoadInt64(&groupsReady[i].incompleteCount)
+		if incompleteCount > 0 || (incompleteCount != dependencyFreeGroup && !completed[i]) {
+			select {
+			case done := <-groupDoneCh:
+				if done.err != nil {
+					// if there is an error, report the error to the output channel.
+					p.outChan <- loadedTransactionGroup{
```

Comment:
should we also set a failed group index to let the caller produce a better error?